pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go

```diff
@@ -0,0 +1,477 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package mcsapi
+
+import (
+	"context"
+	"log/slog"
+	"maps"
+	"slices"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	mcsapiv1alpha1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+
+	controllerruntime "github.com/cilium/cilium/operator/pkg/controller-runtime"
+	"github.com/cilium/cilium/pkg/k8s/utils"
+)
+
+const (
+	// endpointSliceLocalMCSAPIControllerName is a unique value used with LabelManagedBy to indicate
+	// the component managing an EndpointSlice.
+	endpointSliceLocalMCSAPIControllerName = "endpointslice-local-mcsapi-controller.cilium.io"
+)
+
+// mcsAPIEndpointSliceMirrorReconciler is a controller that mirrors local
+// EndpointSlice from a local Service with a ServiceExport to its derived Service.
+// It works by mirroring the EndpointSlice by its suffix name (the name without
+// the generatedName prefix).
+type mcsAPIEndpointSliceMirrorReconciler struct {
+	client.Client
+	Logger *slog.Logger
+
+	clusterName string
+}
+
+func newMCSAPIEndpointSliceMirrorReconciler(mgr ctrl.Manager, logger *slog.Logger, clusterName string) *mcsAPIEndpointSliceMirrorReconciler {
+	return &mcsAPIEndpointSliceMirrorReconciler{
+		Client:      mgr.GetClient(),
+		Logger:      logger,
+		clusterName: clusterName,
+	}
+}
+
+func (r *mcsAPIEndpointSliceMirrorReconciler) getDerivedService(ctx context.Context, key types.NamespacedName) (*corev1.Service, error) {
+	var derivedService corev1.Service
+	if err := r.Client.Get(ctx, key, &derivedService); err != nil {
+		return nil, client.IgnoreNotFound(err)
+	}
+	return &derivedService, nil
+}
+
+func (r *mcsAPIEndpointSliceMirrorReconciler) getLocalEndpointSlice(
+	ctx context.Context, derivedEpSlice *discoveryv1.EndpointSlice, derivedService *corev1.Service,
+) (*discoveryv1.EndpointSlice, error) {
+	serviceName := derivedEpSlice.Labels[mcsapiv1alpha1.LabelServiceName]
+	if serviceName == "" {
+		return nil, nil
+	}
+	suffix := getSuffix(derivedEpSlice)
+	var localEpSlice discoveryv1.EndpointSlice
+	if err := r.Client.Get(
+		ctx,
+		types.NamespacedName{Name: serviceName + "-" + suffix, Namespace: derivedEpSlice.Namespace},
```

Comment:
This breaks if the local epslice is not named after the service name. A more robust approach could be having an annotation in the derived epslice with the name of the corresponding original one.